### PR TITLE
crimson/osd, RFC: fix premature retirement of client req while sending OSDMap

### PR DESCRIPTION
**Pull request by rzarzynski**
We need to go through the entire pipeline even when early exiting. Otherwise we'll brake the ordering and blow up `OpSequencer`.

Similiar problem affects in the `is_pg_op() == true` branch which will be fixed in a follow-up commit if this one works as expected.

Likely the `EIO` handling must be revised as well.

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -68,6 +68,39 @@ void ClientRequest::may_set_prev_op()
   }
 }
 
+ClientRequest::interruptible_future<> ClientRequest::send_incremental_map(PG& pg)
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/client_request.cc:71]: I don't think any of this is necessary -- there's no ordering relationship at all between maps sent by the osd to catch up the client and IO requests.

Comment (rzarzynski) [reply src/crimson/osd/osd_operations/client_request.cc:71]: Thanks! This strips one of two two reasons for maintaining the order: the externally, client-visible order-of-responses.

However, there is still the second, internal one: the implicit contract between `ClientRequest` and `OpSequencer`. The latter expects `finish_op()` to be always called in the order.

Comment (athanatos) [reply src/crimson/osd/osd_operations/client_request.cc:71]: Sending the incremental map should really be independent of the client request response, they really aren't particularly related.

Comment (tchaikov) [reply src/crimson/osd/osd_operations/client_request.cc:71]: the contract is fixed by #41882.

Comment (rzarzynski) [reply src/crimson/osd/osd_operations/client_request.cc:71]: Nah, I think it's about a different thing: handling PG-targeting `MOSDOp`.

Also, there are a few other early exits in the `ClientRequest`'s flow as well as the recovery-on-`EIO`.

Comment (tchaikov) [reply src/crimson/osd/osd_operations/client_request.cc:71]: ahh, right! we also need to ensure that they behave in those cases.

**Pull request closed without merging**